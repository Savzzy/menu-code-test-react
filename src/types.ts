import { MenuItemType } from "./components/Menu/components/MenuCategory";

export enum Diners {
  diner1 = "diner 1",
  diner2 = "diner 2",
}

export interface OrderedItemType {
  id: number;
  name: string;
  price: number;
  category: string;
}

export interface Order {
  [key: string]: {
    twoItemsAdded: boolean;
    mainsAdded: boolean;
    orderedItems: Array<OrderedItemType>;
  };
}

export interface Store {
  order: Order;
  activeDiner: string;
  error?: string;
}

export interface UnsplashResponse {
  results: Array<{
    urls: {
      thumb: string;
    };
  }>;
}

//#region action payload types

export interface ItemToAdd {
  menuItem: MenuItemType;
  menuCategory: string;
}

export interface ItemToRemove {
  id: number;
}

export interface ChangeActiveDiner {
  activeDiner: Diners;
}

export interface UpdateErrorState {
  error: string;
}

export type PayloadTypes =
  | ItemToAdd
  | ItemToRemove
  | ChangeActiveDiner
  | UpdateErrorState;

//#endregion

//#region action types

interface ActionType {
  type: string;
}

export interface AddDishToOrderActionType extends ActionType {
  payload: ItemToAdd;
}

export interface RemoveDishFromOrderActionType extends ActionType {
  payload: ItemToRemove;
}

export interface ChangeActiveDinerActionType extends ActionType {
  payload: ChangeActiveDiner;
}

export interface UpdateErrorStateActionType extends ActionType {
  payload: UpdateErrorState;
}

export type ActionTypes =
  | AddDishToOrderActionType
  | RemoveDishFromOrderActionType
  | ChangeActiveDinerActionType
  | UpdateErrorStateActionType;

//#endregion
