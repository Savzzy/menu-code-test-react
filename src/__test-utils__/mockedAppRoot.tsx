import { render, RenderResult } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { ThemeProvider } from "styled-components";
import globalTheme from "../globalTheme";
import rootReducer from "../reducers";
import { intialStore } from "../store";
import { Store } from "../types";

interface MockedAppRootOptions {
  mockStore: Store;
}

const mockedAppRoot = (
  ComponentTree: JSX.Element,
  options?: MockedAppRootOptions,
): RenderResult => {
  const AppContainer: React.FC = ({ children }) => (
    <Provider
      store={createStore(rootReducer, options?.mockStore || intialStore)}
    >
      <ThemeProvider theme={globalTheme}>{children}</ThemeProvider>
    </Provider>
  );

  return render(ComponentTree, { wrapper: AppContainer });
};

export default mockedAppRoot;
