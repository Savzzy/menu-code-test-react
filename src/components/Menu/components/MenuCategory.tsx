import equal from "fast-deep-equal";
import React, { useEffect, useRef } from "react";
import { Col, Grid, Row } from "react-flexbox-grid";
import { useDispatch } from "react-redux";
import styled from "styled-components";
import { addItemToOrder } from "../../../actions";
import MenuItem from "./MenuItem";

const CategoryName = styled.h2`
  color: ${(props) => props.theme.colors.primaryText};
  font-size: 20px;
  margin: 25px 0 20px 20px;
  text-transform: capitalize;
  font-weight: 400;
`;

const MenuItemsContainer = styled(Grid)``;

export interface MenuItemType {
  id: number;
  name: string;
  price: number;
}

interface MenuCategoryProps {
  menuCategory: string;
  menuItems: Array<MenuItemType>;
  registerOffset?: (menuCategory: string, offsetY: number) => void;
}

const MenuCategory: React.FC<MenuCategoryProps> = ({
  menuCategory,
  menuItems,
  registerOffset,
}) => {
  const categoryName = useRef(null);

  useEffect(() => {
    if (categoryName?.current && registerOffset) {
      registerOffset(
        menuCategory,
        categoryName.current.getBoundingClientRect().y,
      );
    }
  });

  const dispatch = useDispatch();

  return (
    <React.Fragment>
      <CategoryName ref={categoryName}>{menuCategory}</CategoryName>
      <MenuItemsContainer fluid>
        <Row>
          {menuItems.map((menuItem: MenuItemType) => {
            return (
              <Col xs={12} sm={6} md={4} key={menuItem.id}>
                <MenuItem
                  key={menuItem.id}
                  menuItem={menuItem}
                  priceUnit="Â£"
                  onClick={(menuItem: MenuItemType) => {
                    dispatch(addItemToOrder(menuItem, menuCategory));
                  }}
                />
              </Col>
            );
          })}
        </Row>
      </MenuItemsContainer>
    </React.Fragment>
  );
};

export default React.memo(MenuCategory, (prevProps, nextProps) => {
  return !equal(prevProps, nextProps);
});
