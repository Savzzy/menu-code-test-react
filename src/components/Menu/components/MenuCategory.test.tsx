import React from "react";
import mockedAppRoot from "../../../__test-utils__/mockedAppRoot";
import MenuCategory, { MenuItemType } from "./MenuCategory";
import {
  mockMenuCategory,
  mockMenuItems,
  mockUnsplashResponse,
} from "./MenuCategory.fixtures";

import axios from "axios";
import { UnsplashResponse } from "../../../types";
import { waitFor } from "@testing-library/react";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe("MenuCategory", () => {
  it("renders correctly", () => {
    const { container } = mockedAppRoot(
      <MenuCategory
        menuCategory={mockMenuCategory}
        menuItems={mockMenuItems}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows the passed category", async () => {
    const { getByText } = mockedAppRoot(
      <MenuCategory
        menuCategory={mockMenuCategory}
        menuItems={mockMenuItems}
      />,
    );

    await waitFor(() => {
      expect(getByText(mockMenuCategory)).toBeInTheDocument();
    });
  });

  it("renders a menu item for each item passed", async () => {
    const resp = { data: mockUnsplashResponse };
    mockedAxios.get.mockResolvedValue(resp);

    const { getAllByTestId, getByText } = mockedAppRoot(
      <MenuCategory
        menuCategory={mockMenuCategory}
        menuItems={mockMenuItems}
      />,
    );

    await waitFor(() => {
      expect(getAllByTestId("menu_item").length).toBe(mockMenuItems.length);
      mockMenuItems.forEach((menuItem: MenuItemType) => {
        expect(getByText(menuItem.name)).toBeInTheDocument();
        expect(getByText(`Â£ ${menuItem.price.toFixed(2)}`)).toBeInTheDocument();
      });
    });
  });
});
