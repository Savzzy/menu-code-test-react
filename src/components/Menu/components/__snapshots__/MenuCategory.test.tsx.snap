// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuCategory renders correctly 1`] = `
<h2
  class="sc-AxmLO ebklRv"
>
  mains
</h2>
`;
